blockchain/transaction_confirmers.go

```diff
@@ -298,36 +299,84 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
 	for {
 		select {
-		case err := <-subscription.Err():
-			log.Error().Err(err).Msg("Error while subscribed to new headers, restarting subscription")
+		case err := <-subscription.Err(): // Most subscription errors are temporary RPC downtime, so let's poll to resubscribe
+			log.Error().Err(err).Msg("Error while subscribed to new headers, likely RPC errors")
 			subscription.Unsubscribe()
 
-			subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
-			if err != nil {
-				log.Error().Err(err).Msg("Failed to resubscribe to new headers")
-				return err
+			rpcDegradedTime, rpcDegradedNotifyTime := time.Now(), time.Now()
+			timeout, ticker := time.NewTimer(e.NetworkConfig.Timeout.Duration), time.NewTicker(time.Second)
+		reSubLoop:
+			for { // Loop to resubscribe to new headers,
+				select {
+				case <-timeout.C: // Timeout waiting for RPC to come back up
+					log.Error().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC connection still down, timed out waiting for it to come back up")
+					e.Client.Close()
+					return
+				case <-ticker.C: // Poll the RPC connection to see if it's back up
+					if time.Since(rpcDegradedNotifyTime) >= time.Second*15 { // Periodically inform that we're still waiting for RPC to come back up
+						log.Warn().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC connection still down, waiting for it to come back up")
+						rpcDegradedNotifyTime = time.Now()
+					}
+					subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
+					if err == nil { // No error on resubscription, RPC connection restored, back to regularly scheduled programming
+						ticker.Stop()
+						log.Info().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC and subscription connection restored")
+						err = e.backfillMissedBlocks(lastHeaderNumber, headerChannel)
+						if err != nil {
+							log.Error().Err(err).Msg("Error backfilling missed blocks, subscriptions may be out of sync")
+						}
+						break reSubLoop
+					}
+					log.Trace().Err(err).Msg("Error trying to resubscribe to new headers, likely RPC down")
+				}
 			}
 		case header := <-headerChannel:
+			lastHeaderNumber = header.Number.Uint64()
 			e.receiveHeader(header)
 		case <-e.doneChan:
 			log.Debug().Str("Network", e.NetworkConfig.Name).Msg("Subscription cancelled")
 			e.Client.Close()
-			return nil
+			return
+		}
+	}
+}
+
+// backfillMissedBlocks checks if there are any missed blocks since a bad connection was detected, and if so, backfills them
+// to our header channel
+func (e *EthereumClient) backfillMissedBlocks(lastBlockSeen uint64, headerChannel chan *SafeEVMHeader) error {
+	latestBlockNumber, err := e.LatestBlockNumber(context.Background())
+	if err != nil {
+		return err
+	}
+	if latestBlockNumber <= lastBlockSeen {
+		return nil
+	}
+	for i := lastBlockSeen + 1; i <= latestBlockNumber; i++ {
+		header, err := e.HeaderByNumber(context.Background(), big.NewInt(int64(i)))
```

Comment:
I am a bit worried on repetitive rpc calls. Can we use [SubscribeNewHead](https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L329) ?